pinot-plugins/pinot-stream-ingestion/pinot-kafka-2.0/src/main/java/org/apache/pinot/plugin/stream/kafka20/KafkaStreamMetadataProvider.java

```diff
@@ -63,6 +69,28 @@ public StreamPartitionMsgOffset fetchStreamPartitionOffset(OffsetCriteria offset
         offset =
             _consumer.beginningOffsets(Collections.singletonList(_topicPartition), Duration.ofMillis(timeoutMillis))
                 .get(_topicPartition);
+      } else if (offsetCriteria.isPeriod()) {
+        String offsetString = offsetCriteria.getOffsetString();
+        Long periodToMillis = TimeUtils.convertPeriodToMillis(offsetString);
+        TopicPartition tp = new TopicPartition(_topic, _partition);
+        _offsetsForTimes.put(tp, System.currentTimeMillis() - periodToMillis);
```

Comment:
Please use an injected `java.time.Clock` so the time can be controlled for testing purposes. Injecting `Clock.systemClock()` and calling `Clock.millis()` is equivalent to this logic, but allows you to inject a fixed clock for testing purposes.